unixfs/src/dir/builder.rs

```diff
@@ -0,0 +1,834 @@
+        })
+    }
+
+    /// Directories get "put" implicitly through the put files, and directories need to be adjusted
+    /// only when wanting them to have metadata.
+    pub fn set_metadata(
+        &mut self,
+        full_path: &str,
+        metadata: Metadata,
+    ) -> Result<(), TreeBuildingFailed> {
+        // create all paths along the way
+        //
+        // set if not set, error otherwise? FIXME: doesn't error atm
+        self.modify_with(full_path, |parent, basename, id| {
+            parent
+                .add_or_get_node(basename, id)
+                .map_err(|_| TreeBuildingFailed::LeafAsDirectory(full_path.to_string()))?
+                .set_metadata(metadata);
+            Ok(())
+        })
+    }
+
+    fn modify_with<F>(&mut self, full_path: &str, f: F) -> Result<(), TreeBuildingFailed>
+    where
+        F: FnOnce(&mut DirBuilder, String, &mut Option<u64>) -> Result<(), TreeBuildingFailed>,
+    {
+        // create all paths along the way
+        //
+        // assuming it's ok to split '/' since that cannot be escaped in linux at least
+
+        self.longest_path = full_path.len().max(self.longest_path);
+        let mut remaining = full_path.split('/').enumerate().peekable();
+        let mut dir_builder = &mut self.root_builder;
+
+        // needed to avoid borrowing into the DirBuilder::new calling closure
+        let counter = &mut self.counter;
+
+        while let Some((depth, next)) = remaining.next() {
+            let last = remaining.peek().is_none();
+
+            match (depth, next, last) {
+                // this might need to be accepted in case there is just a single file
+                (0, "", true) => { /* accepted */ }
+                (0, "", false) => {
+                    return Err(TreeBuildingFailed::RootedPath(full_path.to_string()))
+                }
+                (_, "", false) => {
+                    return Err(TreeBuildingFailed::RepeatSlashesInPath(
+                        full_path.to_string(),
+                    ))
+                }
+                (_, "", true) => todo!("path ends in slash"),
+                _ => {}
+            }
+
+            // our first level can be full given the options
+            let full = depth == 0 && !self.opts.wrap_in_directory && dir_builder.is_empty();
+
+            if last {
+                let mut next_id = Some(*counter);
+
+                let ret = if full {
+                    Err(TreeBuildingFailed::TooManyRootLevelEntries)
+                } else {
+                    f(dir_builder, next.to_string(), &mut next_id)
+                };
+
+                if next_id.is_none() {
+                    *counter += 1;
+                }
+
+                if ret.is_err() {
+                    // FIXME: there might be a case where we have now stale nodes in our tree but
+                    // cannot figure out an example for that.
```

Comment:
this is definitely a point to think about to "do something" in a longer path which fails will not create new nodes (it can only fail for using leaf as a dir or duplicate name so it's stopped early before creating new nodes). here "do something" includes both ops: adding a file and setting metadata (for a directory).

files cannot have metadata being set as the file tree itself would have this information. Technically this could be supported and would be a valid usecase; the root of the file would then just need to be rewritten.